Add optimal thresholds export function

Maya084 opened the pull request:
This PR adds a function that exports the optimal confidence thresholds for each class in a json and csv format. 

Resolves #5921

I have read the CLA Document and I hereby sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds automatic export of optimal thresholds and maximum F1 scores for each class during model evaluation, saving results as CSV and JSON files. 📈📝

### 📊 Key Changes
- Introduced a new `export_optimal_thresholds` function to compute and save the best threshold and F1 score per class.
- Updated the `ap_per_class` function to call this export function after evaluation.
- Documentation updated to reference the new function.
- Minor update to author metadata.

### 🎯 Purpose & Impact
- Makes it easier for users to find the best confidence thresholds for each class, improving model deployment and performance tuning.
- Results are saved in both CSV and JSON formats, supporting easy integration with other tools and workflows.
- Enhances transparency and reproducibility in model evaluation.

Maya084 committed:
Add optimal thresholds export function

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Merge branch 'main' into optimal_thresholds

Maya084 commented:
I have read the CLA Document and I sign the CLA

glenn-jocher commented:
@Maya084 wow this is really cool! Can you add a call to LOGGER.info() to print the values to the console for viewing also?

glenn-jocher committed:
Merge branch 'main' into optimal_thresholds

Maya084 commented:
@glenn-jocher Hey Glenn, I'm glad you liked the pull request! I've added the logging as you suggested, but I encountered an issue with the order of output in the console. The logging values appear before the output from the validation function, resulting in a less-than-ideal console display.
Here's an example of how it currently looks (I also added values for precision and recall at the optimal threshold):

```
val: Scanning /folder/labels.cache... 100 images, 0 backgrounds, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]
                 Class       Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 7/7 [00:02<00:00,  2.44it/s]
                    Class    Optimal Threshold         F1  Precision     Recall
               class_1                0.604        1.0        1.0        1.0
               class_2                 0.69        1.0        1.0        1.0
               class_3                0.746      0.923        1.0      0.857
               class_4                0.831        1.0        1.0        1.0
               class_5                  0.3        1.0        1.0        1.0
               class_6                0.762        0.8        1.0      0.667
               class_7                0.253      0.856      0.748        1.0
               class_8                0.126      0.923      0.857        1.0
               class_9                0.603      0.933        1.0      0.874
               class_10               0.206      0.667       0.75        0.6
            all        100        185      0.785      0.839      0.901      0.764
        class_1        100          5      0.929          1      0.995      0.662
        class_2        100          9        0.8          1      0.995      0.922
        class_3        100          7      0.799      0.857      0.964      0.904
        class_4        100          2      0.694          1      0.995      0.995
        class_5        100         11          1      0.956      0.995      0.854
        class_6        100          6      0.854      0.667      0.815      0.662
        class_7        100          6      0.819      0.762      0.927      0.612
Speed: 2.4ms preprocess, 12.7ms inference, 0.0ms loss, 1.3ms postprocess per image
```

As you can see, the header from the final results is before the new output.
I'm not entirely satisfied with this, as it might make it harder to follow the output. If it's acceptable to you, I can commit the changes as they are. However, I wanted to bring this to your attention in case you have any recommendations on how to improve the logging placement. I'm not fully familiar with the logging process, and I couldn't easily find a way to ensure that the logging of the optimal thresholds appears after everything else is printed.

glenn-jocher commented:
@Maya084 thank you for the contribution! The addition of the `export_optimal_thresholds` function to export optimal confidence thresholds and max F1 scores is a valuable enhancement. The availability of results in both CSV and JSON formats will indeed make it easier for users to analyze and integrate these metrics into their workflows. This feature should help in fine-tuning the decision-making process for each class, potentially leading to better model performance when deployed in real-world scenarios. 🎉

We appreciate your efforts in improving the YOLOv8 repository and look forward to reviewing your PR. If you have any further suggestions or need assistance, feel free to reach out. 👍

glenn-jocher committed:
Merge branch 'main' into optimal_thresholds

Maya084 committed:
rename csv header

glenn-jocher committed:
Merge branch 'main' into optimal_thresholds

glenn-jocher committed:
Merge branch 'main' into optimal_thresholds

glenn-jocher committed:
Merge branch 'main' into optimal_thresholds

Burhan-Q commented:
@Maya084 I think there are a few places that would need to get updated to correct the formatting for the `LOGGER`. One place (but would also be for all tasks) is in the [val.py](https://github.com/ultralytics/ultralytics/blob/2f22ff4ae66e86c53def72d52eb2e61ec924172e/ultralytics/models/yolo/detect/val.py#L178) (this is only for `detect` as an example). Since this is where the formatting for the `LOGGER` output would be. Additionally you'd need to ensure that you have the headers updated, which I believe would be [here](https://github.com/ultralytics/ultralytics/blob/2f22ff4ae66e86c53def72d52eb2e61ec924172e/ultralytics/utils/metrics.py#L854). I'm not 100% certain, but those might be good places to check to fix the formatting, which I suspect will be necessary.

UltralyticsAssistant committed:
Merge branch 'main' into optimal_thresholds

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into optimal_thresholds

UltralyticsAssistant committed:
Merge branch 'main' into optimal_thresholds

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions
